cpp/src/neighbors/all_neighbors/all_neighbors_merge.cuh

```diff
@@ -0,0 +1,266 @@
+                                   float* batch_distances,
+                                   IdxT* global_indices,
+                                   IdxT* batch_indices,
+                                   bool select_min)
+{
+  size_t batch_row = blockIdx.x;
+  typedef cub::BlockMergeSort<raft::KeyValuePair<float, IdxT>, BLOCK_SIZE, ITEMS_PER_THREAD>
+    BlockMergeSortType;
+  __shared__ typename cub::BlockMergeSort<raft::KeyValuePair<float, IdxT>,
+                                          BLOCK_SIZE,
+                                          ITEMS_PER_THREAD>::TempStorage tmpSmem;
+
+  extern __shared__ char sharedMem[];
+  float* blockKeys  = reinterpret_cast<float*>(sharedMem);
+  IdxT* blockValues = reinterpret_cast<IdxT*>(&sharedMem[graph_degree * 2 * sizeof(float)]);
+  int16_t* uniqueMask =
+    reinterpret_cast<int16_t*>(&sharedMem[graph_degree * 2 * (sizeof(float) + sizeof(IdxT))]);
+
+  if (batch_row < num_cluster_in_batch) {
+    // load batch or global depending on threadIdx
+    size_t global_row = cluster_data_indices[batch_row];
+
+    raft::KeyValuePair<float, IdxT> threadKeyValuePair[ITEMS_PER_THREAD];
+
+    size_t halfway   = BLOCK_SIZE / 2;
+    size_t do_global = threadIdx.x < halfway;
+
+    float* distances;
+    IdxT* indices;
+
+    if (do_global) {
+      distances = global_distances;
+      indices   = global_indices;
+    } else {
+      distances = batch_distances;
+      indices   = batch_indices;
+    }
+
+    size_t idxBase = (threadIdx.x * do_global + (threadIdx.x - halfway) * (1lu - do_global)) *
+                     static_cast<size_t>(ITEMS_PER_THREAD);
+    size_t arrIdxBase = (global_row * do_global + batch_row * (1lu - do_global)) * graph_degree;
+    for (int i = 0; i < ITEMS_PER_THREAD; i++) {
+      size_t colId = idxBase + i;
+      if (colId < graph_degree) {
+        threadKeyValuePair[i].key   = distances[arrIdxBase + colId];
+        threadKeyValuePair[i].value = indices[arrIdxBase + colId];
+      } else {
+        if (select_min) {
+          threadKeyValuePair[i].key   = std::numeric_limits<float>::max();
+          threadKeyValuePair[i].value = std::numeric_limits<IdxT>::max();
+        } else {
+          threadKeyValuePair[i].key   = std::numeric_limits<float>::min();
+          threadKeyValuePair[i].value = std::numeric_limits<IdxT>::min();
+        }
+      }
+    }
+
+    __syncthreads();
+
+    if (select_min) {
+      BlockMergeSortType(tmpSmem).Sort(threadKeyValuePair, raft::less_op{});
+    } else {
+      BlockMergeSortType(tmpSmem).Sort(threadKeyValuePair, raft::greater_op{});
+    }
+
+    // load sorted result into shared memory to get unique values
+    idxBase = threadIdx.x * ITEMS_PER_THREAD;
+    for (int i = 0; i < ITEMS_PER_THREAD; i++) {
+      size_t colId = idxBase + i;
+      if (colId < 2 * graph_degree) {
+        blockKeys[colId]   = threadKeyValuePair[i].key;
+        blockValues[colId] = threadKeyValuePair[i].value;
+      }
+    }
+
+    __syncthreads();
```

Comment:
Correct my understanding if I am wrong, but shouldn't this `__syncthreads` be after the `Sort` operation (before storing the sorted stuff in `blockKeys` and `blockValues`)? Also, for computing the `uniqueMask`, it seems each thread is handling the values written by itself, so we might be able to avoid this sync after the storing.

Reply:
Ok the `uniqueMask` makes sense. But the reason I ask about the `__syncthreads` after the `Sort` is that we do it that way in Vamana (the `Sort` is preceded and succeeded by `__syncthreads`). If you are sure that the `BlockMergeSort` synchronizes before returning, this is fine. I don't want to go too deep into the implementation details here, just making sure we don't introduce a bug.

Reply:
Thanks for pointing out. After looking up a bit it looks like there is no _explicitly_ guarantee of a sync. I'll add syncthreads to the code, shouldn't hurt to be safe.

Reply:
Yes, I looked up some other examples too, which do a `__syncthreads` both before and after `BlockMergeSort` / `BlockRadixSort`. We should be consistent.